### PR TITLE
Write an internet draft

### LINKED ISSUE #71: Start an internet-draft
This supports the [IETF submission](https://mailarchive.ietf.org/arch/msg/dispatch/NQ0deHSsRvt4BL4alk_WYVnhhvo). It has to happen before they close the RFC database from July 3 until the [Prague meeting](http://ietf.org/meeting/99/index.html).

We can use https://github.com/cabo/kramdown-rfc2629 to author in markdown, and https://github.com/martinthomson/i-d-template to build with Travis CI.

### PR DESCRIPTION
**Pull request by jyasskin**
Fixes #71. I'm going to submit the draft to the IETF before this is reviewed in order to make their July 3 deadline, but I definitely want y'all to point out problems so I can fix them in the next draft.

I've pre-emptively removed the :method header (#70) and changed the `sections` map into an array, and I'll backfill those into the explainer and implementation when I get a bit more time.

### CODE REVIEW

File: draft-yasskin-dispatch-web-packaging.md [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+1. The bytes of the `manifest` CBOR item.
+
+Let *signing-certificates* be an empty array.
+
+For each element *signature* of `signatures`:
+
+1. Let *certificate* be `certificates`\[*signature*\["keyIndex"]].
+
+1. The parser MUST define a partial function from public key types to signing
+   algorithms, with the following map as a subset:
```
Comment (mrdewitt) [inline draft-yasskin-dispatch-web-packaging.md]: "as a subset" isn't very clear to me, perhaps this is mathematician lingo though?

File: draft-yasskin-dispatch-web-packaging.md [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+   continue to the next *signature*.
+
+1. Append *certificate* to *signing-certificates*. Note that failed signatures
+   simply cause their certificate to be ignored, so that packagers can give new
+   signature types to parsers that understand them.
+
+
+Let *origin* be `manifest`\["metadata"]\["origin"].
+
+Try to find a certificate in *signing-certificates* that has an identity
```
Comment (mrdewitt) [inline draft-yasskin-dispatch-web-packaging.md]: "Try to find" is unclear, perhaps something like "iterate through signing-certificates until one is found that has an identity matching ...

File: draft-yasskin-dispatch-web-packaging.md [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+
+
+### Sub-packages {#subpackages}
+
+A sub-package is represented by a [manifest](#manifest) file looked up as
+a [resource](#resource) within the `indexed-content` section. The sub-package's
+resources are not otherwise distinguished from the rest of the resources in the
+package. Sub-packages can form an arbitrarily-deep tree.
+
+There are three possible forms of dependencies on sub-packages, of which we
```
Comment (mrdewitt) [inline draft-yasskin-dispatch-web-packaging.md]: Perhaps reorder this paragraph after the cddl defining a subpackage?

File: draft-yasskin-dispatch-web-packaging.md [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+
+~~~~~ cbor-diag
+[32("https://example.com/loginsdk.package"),
+ {"hash": {"sha256":
+     b64'9qg0NGDuhsjeGwrcbaxMKZAvfzAHJ2d8L7NkDzXhgHk='}}]
+~~~~~
+
+Note that because the sub-package may include sub-sub-packages by date, the top
+package may need to explicitly list those sub-sub-packages' hashes in order to
+be completely constrained.
```
Comment (mrdewitt) [inline draft-yasskin-dispatch-web-packaging.md]: nit: Could use an example.

File: draft-yasskin-dispatch-web-packaging.md [hunk 5]
```diff
@@ -0,0 +1,650 @@
...
+* the private key for any certificate in the signature's validation
+  chain is leaked, or
+* a vulnerability is discovered in the package's contents.
+
+Because packages are intended to be used offline, it's impossible to
+inject a revocation check into the critical path of using the package,
+and even in online scenarios,
+such
+[revocation checks don't actually work](https://www.imperialviolet.org/2012/02/05/crlsets.html).
+Instead, package consumers must check for a sufficiently recent set of
```
Comment (mrdewitt) [inline draft-yasskin-dispatch-web-packaging.md]: Does this mean that packages become unusable after 7-30 days offline?

File: draft-yasskin-dispatch-web-packaging.md [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+keyword: Internet-Draft
+
+stand_alone: yes
+pi: [comments, sortrefs, strict, symrefs, toc]
+
+author:
+ -
+    name: Jeffrey Yasskin
+    organization: Google
+    email: jyasskin@chromium.org
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Folks seem to vary. Several people use @google, but https://tools.ietf.org/html/rfc6797 uses a personal address.

File: draft-yasskin-dispatch-web-packaging.md [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+details can be found in that repository.
+
+Use Cases    {#use-cases}
+---------
+
+### Offline Installation
+
+People with expensive or intermittent internet connections are used
+to sharing files via P2P links and shared SD cards. They should be
+able to install web applications they received this way. Installing a
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: I don't know a well-defined term for this. "Install" doesn't work well with "websites", which also isn't particularly defined. "Service workers" have a defined install step, but it's not obvious that we want to insist on service workers for these packages. Any suggestions?

File: draft-yasskin-dispatch-web-packaging.md [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+Why not ZIP?   {#not-zip}
+------------
+
+[WICG/webpackage#45](https://github.com/WICG/webpackage/issues/45)
+
+The Need for Standardization   {#need}
+----------------------------
+
+Publishers and readers should be able to generate a package once, and have it
+usable by all browsers.
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: That's true, but I'm not sure it addresses the need for standardization?

File: draft-yasskin-dispatch-web-packaging.md [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+
+This specification creates local variables with the phrase "Let *variable-name*
+be ...". Use of a variable before it's created is a defect in this
+specification.
+
+## Top-level structure  {#top-level}
+
+The package is roughly a CBOR item with the following CDDL schema, but package
+parsers are required to successfully parse some byte strings that aren't valid
+CBOR. For example, sections may have padding between them, or even overlap, as
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: We can't prohibit overlap without forcing extra work onto parsers.

File: draft-yasskin-dispatch-web-packaging.md [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+1. Any `resource-key`'s first three headers are not named ":scheme",
+   ":authority", and ":path", in that order. Note that ":method" is
+   intentionally omitted because only the GET method is meaningful.
+1. Any of the pseudo-headers' values violates a requirement in Section
+   8.1.2.3 of {{!HTTP2=RFC7540}}.
+1. Any `resource-key` has a non-pseudo-header name that includes the
+   ":" character or is not lower-case ascii ({{HTTP2}}, Section
+   8.1.2).
+1. Any two decoded `resource-key`s are the same. Note that header
+   lists with the same header fields in a different order are not the
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Yes! That'll appear in the W3C spec about browsers loading packages, but basically, we're not going to (can't) locate resources by looking for an exact match between the headers the browser would generate and the key in the index. Instead, the browser will have to use its request information to pick the best of the options provided by the package.

File: draft-yasskin-dispatch-web-packaging.md (hunk 5 above)
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Yes, although there'll be a cheap way to update them and to distribute those updates offline. I can make an argument that we should let a package live forever on a given device as long as that device _never_ goes online, but I'm not sure it'll convince everyone.

File: draft-yasskin-dispatch-web-packaging.md (hunk 4 above)
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Done.

File: draft-yasskin-dispatch-web-packaging.md (hunk 3 above)
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Done.

File: draft-yasskin-dispatch-web-packaging.md (hunk 2 above)
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Great, thanks.

File: draft-yasskin-dispatch-web-packaging.md (hunk 1 above)
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Yes, derived from a function being a particular kind of [relation](https://en.wikipedia.org/wiki/Binary_relation), which is a set of pairs; but this document should be comprehensible to non-mathematicians, so I've changed it.

File: draft-yasskin-dispatch-web-packaging.md [hunk 11]
```diff
@@ -0,0 +1,650 @@
...
+
+Load a CBOR item starting at *index-start* + 1 as the `index` array in the following CDDL:
+
+~~~~~ cddl
+index = [* [resource-key: http-headers,
+            offset: uint,
+            ? length: uint] ]
+
+; http-headers is a byte string in HPACK format (RFC7541).
+; The dynamic table begins empty for each instance of
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Yeah ... How about I delete this comment about the dynamic table and rely on the actual parsing described below.

File: draft-yasskin-dispatch-web-packaging.md [hunk 12]
```diff
@@ -0,0 +1,650 @@
...
+
+Use Cases    {#use-cases}
+---------
+
+### Offline Installation
+
+People with expensive or intermittent internet connections are used
+to sharing files via P2P links and shared SD cards. They should be
+able to install web applications they received this way. Installing a
+web application requires a TLS-type guarantee that it came from and
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: How's this?

File: draft-yasskin-dispatch-web-packaging.md [hunk 13]
```diff
@@ -0,0 +1,650 @@
...
+  appmanifest: W3C.WD-appmanifest-20170608
+  SRI: W3C.REC-SRI-20160623
+
+informative:
+  ServiceWorkers: W3C.WD-service-workers-1-20161011
+
+--- abstract
+
+Web Packages provide a way to bundle up groups of web resources to
+transmit them together. These bundles can then be signed to establish
```
Comment (jyasskin) [inline draft-yasskin-dispatch-web-packaging.md]: Done.

**Comment by jyasskin** [comment]
I'm going to merge this to make it easier for me to build on top of it. Feel free to send other comments or file bugs about what needs to change.

**Pull request merged**